### PR TITLE
Update BPipeline to handle missing data

### PR DESCRIPTION
**Pull request by amishatishpatel**
What initially started as an update for the BPipeline crept a bit into a feature tweak
on the XPipeline too. This was all to accommodate more robust data testing/handling
in the end-to-end unit test; specifically to test non-contiguous data (e.g. missed Chunks
in the receiver).

The unit test update did become quite an adventure as mypy eventually wasn't happy
with how complex it had become - hence, some abstractions. I did try to explain the 
unit test updates as much, and as often, as possible. As always, let me know if it's just
more confusing/needs tweaking.

Just realised I have tweaked an existing design of the XPipeline. Should I lump that under
the "update XBEngine documentation" ticket?

Checklist (if not applicable, edit to add `(N/A)` and mark as done):

- [x] (N/A) If dependencies are added/removed: update `setup.cfg` and `.pre-commit-config.yaml`
- [x] (N/A) If modules are added/removed: use `sphinx-apidoc -efo doc/ src/` to update files in `doc/`
- [x] Ensure copyright notices are present and up-to-date
- [x] (N/A) If qualification tests are changed: attach a sample qualification report
- [ ] If design has changed: ensure documentation is up to date
- [x] (N/A) If ICD-defined sensors have been added: update `fake_servers.py` in katsdpcontroller to match

Closes NGC-1172.

### CODE REVIEW

**Comment by bmerry** [comment]
@amishatishpatel we'll also need to update https://docs.google.com/document/d/14irzt0m5SOtKDWFzZG6xYbwhWuLQK52QIHFDaCqrCk0/edit. Do you want to do that under NGC-1172 or a new ticket?

**Comment by amishatishpatel** [comment]
> @amishatishpatel we'll also need to update https://docs.google.com/document/d/14irzt0m5SOtKDWFzZG6xYbwhWuLQK52QIHFDaCqrCk0/edit. Do you want to do that under NGC-1172 or a new ticket?

Argh, too right. Admittedly the first I'm seeing/hearing of such an officially unofficial document. Still
* It should definitely be part of NGC-1172 (so I/we don't forget). 
* Maybe I'll create a child task.

**Review by bmerry — CHANGES_REQUESTED** [review]
Some changes you can start with on the implementation. I'll look at the test updates next.

File: test/xbgpu/test_engine.py
Comment (bmerry) [inline test/xbgpu/test_engine.py:1021]: Can you be sure that they come in this order?

If not, I think caplog has facilities to ignore particular warnings for the duration of the test.

File: test/xbgpu/test_engine.py
Comment (bmerry) [inline test/xbgpu/test_engine.py:1053]: I would suggest modifying generate_expected_beams to take a list of batch indices to generate, instead of a range, so that you don't need to do this nested looping. It's an `njit` function and those don't cope that well with Python lists, so you should convert the list to an ndarray when passing it in.

**Review by bmerry — CHANGES_REQUESTED** [review]
Here are some initial review comments on the tests. I'm still trying to get my head around all the test changes. One concern I have is that the choice of batch indices to send still revolves around the accumulation interval, and doesn't consider heaps_per_fengine_per_chunk. Ideally I'd like to ensure that we have full, partial and skipped chunks of input to ensure that those cases are all handled. Maybe that's already the case though? I haven't worked through the numbers.

File: src/katgpucbf/xbgpu/engine.py
Comment (amishatishpatel) [reply src/katgpucbf/xbgpu/engine.py:155]: Admittedly in a prison of my own making. Would you be ok if I updated this to refer to the Chunk in a general sense, e.g.
- "[...] of which antennas have missed data in the **current received Chunk** being processed."
  - Similar to the docstring for `RxQueueItem`

There's probably a reason why I can't find references to files/classes from the `src/katgpucbf/` level...

Comment (bmerry) [reply src/katgpucbf/xbgpu/engine.py:155]: Sphinx lets you easily reference things from the current module/class, but it doesn't do relative paths in general. If you want a link, fully qualify it e.g. ``:class:`~katgpucbf.fgpu.recv.Chunk` `` (the leading `~` makes it render just as `Chunk`).

Comment (amishatishpatel) [reply src/katgpucbf/xbgpu/engine.py:155]: Yeah I figured I could fully spell it out for Sphinx - for some reason that felt a bit dirty to me. Still, can't argue if it works; will update accordingly.

File: src/katgpucbf/xbgpu/engine.py
Comment (amishatishpatel) [reply src/katgpucbf/xbgpu/engine.py:508]: This must be a relic of me debugging why the `beam_ants` SPEAD item wasn't updating during dev/testing. I threw all combination of methods at it to ensure I wasn't unintentionally linking arrays with incorrect assignment.

Noted and noted Re:
> That's going to make a temporary copy, which you then copy again into `tx_item.present`.

File: test/xbgpu/test_bsend.py
Comment (amishatishpatel) [reply test/xbgpu/test_bsend.py:124]: I appreciate your diplomacy 😂

File: test/xbgpu/test_engine.py
Comment (amishatishpatel) [reply test/xbgpu/test_engine.py:290]: Yeah genuinely unsure, if I'm honest. Going back through the git blame, can't really find a corresponding check that used to exist before/above it.

I'll remove it now.

File: test/xbgpu/test_engine.py
Comment (amishatishpatel) [reply test/xbgpu/test_engine.py:661]: My apologies, most certainly leftover from reshuffling things.

File: test/xbgpu/test_engine.py
Comment (amishatishpatel) [reply test/xbgpu/test_engine.py:663]: This looks a lot like an iteration of the test I had a while back - tracking heap counts per accumulation.

Appreciate the suggestion, will definitely update it to this sort of MO.

File: test/xbgpu/test_engine.py
Comment (amishatishpatel) [reply test/xbgpu/test_engine.py:1021]: Did some digging this morning and found this ~recent PR on pytest that adds the ability to filter within the unit test
- https://github.com/pytest-dev/pytest/pull/11625/files

I see it's part of a newer version of pytest, which we might not necessarily want to upgrade to yet.

I'll use a plain old `filter` for now, maybe create a technical-debt task for updating version and usage.

Comment (bmerry) [reply test/xbgpu/test_engine.py:1021]: I'd say try updating pytest to the latest and check what happens, but be sure to also test the qualification tests. I vaguely recall that pytest-asyncio 0.23 has some updates to make it compatible with pytest 8, and we're stuck on pytest-asyncio 0.21 until they fix an issue with how scoped fixtures work that breaks the qualification tests.

Comment (amishatishpatel) [reply test/xbgpu/test_engine.py:1017]: I kicked off a qualification run (on the site Jenkins) around 13:15 today with katgpucbf/upgrade-pytest.

Will likely have to revisit this tomorrow morning. I've done a normal `filter` for now.

Comment (amishatishpatel) [reply test/xbgpu/test_engine.py:1017]: The qualification run seems to have passed with no complaints/failures. I'll send you the build URL via Mattermost.

Guess I have a bit more work to do before I can re-request a review...

File: test/xbgpu/test_engine.py
Comment (bmerry) [inline test/xbgpu/test_engine.py:503]:
```suggestion
        return record.message == "All Antennas had a break in data during this accumulation":
```

File: test/xbgpu/test_engine.py
Comment (bmerry) [inline test/xbgpu/test_engine.py:382]:
```suggestion
        # is expected to be incomplete. We also accommodate any other
```

File: test/xbgpu/test_engine.py
Comment (bmerry) [inline test/xbgpu/test_engine.py:1042]: It looks like you've put in the filtering now, and could remove the `: len(corrprod_outputs)`?

**Review by bmerry — CHANGES_REQUESTED** [review]

File: test/xbgpu/test_engine.py
Comment (amishatishpatel) [reply test/xbgpu/test_engine.py:230]: Too right. I wasn't sure what is/isn't supported by `@njit`, so defaulted to a (slightly) more convoluted approach.
* That is, from my reading it seems `numpy`-related operations are more likely supported than native Python operations (?)

File: test/xbgpu/test_engine.py
Comment (amishatishpatel) [reply test/xbgpu/test_engine.py:1055]: Glad you suggested this. I had something vastly similar in a previous iteration of the test (maybe even the previous `HEAD`).

I think I was getting a bit iffy with the amount of things being returned and passed around (also because I'm on two self-imposed strikes with `mypy`'s "test_engine_end_to_end is too complex").

Still, will update accordingly now.

File: test/xbgpu/test_engine.py
Comment (bmerry) [reply test/xbgpu/test_engine.py:230]: Oh, hadn't spotted that it was an `@njit`. But I guess try it and see if it works. If it doesn't work, I'd suggest just having a counter you increment each time around the loop.

File: test/xbgpu/test_engine.py
Comment (amishatishpatel) [reply test/xbgpu/test_engine.py:1042]: Unfortunately not. Some combinations of `heap_accumulation_thresholds` - `(3,7), (5, 9)` - still report the "All Antennas [...]" warning during that jump in data in the middle of the test (the artificial break in data we introduced).
* Those warnings may technically be true, but not exactly because of our decision to start 'in the middle of a Chunk' at the start of the test.

Let me know if that does/doesn't make sense.

File: test/xbgpu/test_engine.py
Comment (amishatishpatel) [inline test/xbgpu/test_engine.py:296]: Hmmmm, maybe this description requires an update. Let me know if you're of the same opinion.

File: test/xbgpu/test_engine.py
Comment (amishatishpatel) [reply test/xbgpu/test_engine.py]: Seems to work fine. I'll chalk this up to another happy accident!

File: src/katgpucbf/xbgpu/engine.py
Comment (bmerry) [inline src/katgpucbf/xbgpu/engine.py:164]:
```suggestion
        An uint32 type :class:`~katsdpsigproc.accel.DeviceArray` with shape
```

File: test/xbgpu/test_engine.py
Comment (bmerry) [reply test/xbgpu/test_engine.py:296]: I agree - my first reaction on seeing the type annotation was "why a list of lists"? Maybe add something like "This is a list of lists, with the outer index matching `corrprod_outputs`."

File: test/xbgpu/test_engine.py
Comment (bmerry) [inline test/xbgpu/test_engine.py:659]:
```suggestion
        acc_indices
```
for consistency with the code.

**Review by bmerry — CHANGES_REQUESTED** [review]
Looking good I think. Before merging I want to check what's going on with the qualification test code I'm developing, which seems to be getting garbage in the SPEAD item. It's probably on my side though (since you are verifying the item in test_bsend).

File: src/katgpucbf/xbgpu/engine.py
Comment (amishatishpatel) [reply src/katgpucbf/xbgpu/engine.py:164]: Looks like a failed phishing attempt atm. Thanks for spotting it.

**Comment by amishatishpatel** [comment]
> Looking good I think. Before merging I want to check what's going on with the qualification test code I'm developing, which seems to be getting garbage in the SPEAD item. It's probably on my side though (since you are verifying the item in test_bsend).

Fair enough. I've pushed changes to address your last round of comments. I'll wait on your go-ahead Re: Merging.

File: test/xbgpu/test_engine.py
Comment (bmerry) [inline test/xbgpu/test_engine.py:298]:
```suggestion
        `corrprod_outputs`. See :meth:`TestEngine.test_engine_end_to_end` for
```

**Review by bmerry — CHANGES_REQUESTED** [review]

File: test/xbgpu/test_engine.py
Comment (bmerry) [inline test/xbgpu/test_engine.py:294]:
```suggestion
        :meth:`TestEngine._send_data`.
```

File: test/xbgpu/test_engine.py
Comment (amishatishpatel) [reply test/xbgpu/test_engine.py:661]: How strange, I committed an updated `test_engine.py` in the previous commit, but this specific change didn't make it through (?). Committing now.

**Review by bmerry — APPROVED** [review]

**Pull request merged**
